simulator/src/ledger/radix_engine_db.rs

```diff
@@ -94,51 +94,54 @@ impl QueryableSubstateStore for RadixEngineDB {
     }
 }
 
-impl SubstateStore for RadixEngineDB {
+impl ReadableSubstateStore for RadixEngineDB {
     fn get_substate<T: Encode>(&self, address: &T) -> Option<Substate> {
         self.read(&scrypto_encode(address))
             .map(|b| scrypto_decode(&b).unwrap())
     }
 
-    fn put_substate<T: Encode>(&mut self, address: &T, substate: Substate) {
-        self.write(&scrypto_encode(address), &scrypto_encode(&substate));
-    }
-
     fn get_child_substate<T: Encode>(&self, address: &T, key: &[u8]) -> Option<Substate> {
         let mut id = scrypto_encode(address);
         id.extend(key.to_vec());
         self.read(&id).map(|b| scrypto_decode(&b).unwrap())
     }
 
-    fn put_child_substate<T: Encode>(&mut self, address: &T, key: &[u8], substate: Substate) {
-        let mut id = scrypto_encode(address);
-        id.extend(key.to_vec());
-        self.write(&id, &scrypto_encode(&substate));
-    }
-
     fn get_epoch(&self) -> u64 {
         let id = scrypto_encode(&"epoch");
         self.read(&id)
             .map(|v| scrypto_decode(&v).unwrap())
             .unwrap_or(0)
     }
 
-    fn set_epoch(&mut self, epoch: u64) {
-        let id = scrypto_encode(&"epoch");
-        let value = scrypto_encode(&epoch);
-        self.write(&id, &value)
-    }
 
     fn get_nonce(&self) -> u64 {
         let id = scrypto_encode(&"nonce");
         self.read(&id)
             .map(|v| scrypto_decode(&v).unwrap())
             .unwrap_or(0)
     }
+}
+
+impl WriteableSubstateStore for RadixEngineDB {
+    fn put_substate(&mut self, address: &[u8], substate: Substate) {
+        self.write(address, &scrypto_encode(&substate));
+    }
+
+    fn put_child_substate(&mut self, address: &[u8], key: &[u8], substate: Substate) {
+        let mut id = address.to_vec();
+        id.extend(key.to_vec());
+        self.write(&id, &scrypto_encode(&substate));
+    }
```

Comment:
Feel like we should get rid of the `put_child_substate` interface. It's causing very inconsistent physical keys in the database.

Reply:
Yup agreed, refining this interface will be the subject of the next few PRs :)